docs/ERROR_HANDLING.md

```diff
@@ -0,0 +1,86 @@
+# Error handling for various entry points
+
+In this document we discuss how different types of errors during contract
+execution are handled by wasmd and the blockchain.
+
+## Two levels of errors
+
+When cosmwasm-vm executes a contract, the caller receives a nested result type:
+`VmResult<ContractResult<R>>` with some success response `R`. The outer
+`VmResult` is created by the host environment and the inner `ContractResult` is
+created inside of the contract. Most application specific error should go into
+`ContractResult` errors. This is what happens when you use `?` inside of your
+contract implementations. The `VmResult`
+[error cases](https://github.com/CosmWasm/cosmwasm/blob/v1.2.3/packages/vm/src/errors/vm_error.rs#L11-L148)
+include e.g.
+
+- Caching errors such as a missing Wasm file or corrupted module
+- Serialization problems in the contract-host communication
+- Panics from panic handler in contract
+- Errors in crypto API calls
+- Out of gas
```

Comment:
Are you sure OutOfGas goes through VM error?
I kind of felt this turns into some other panic n the wasmvm rather than an error. But I may be wrong.. it was a long time since I looked there. Just asking to confirm in the code

Reply:
There are cases in which the VM calls into the chain (e.g. humanize, DB access). Those calls get their own gas meter. When they run out of gas, the error propagates chain -> wasmvm -> cosmwasm-vm -> wasmvm -> chain as a `VmError::GasDepletion`. Also when the Wasmer metering runs out of points, the execution is terminated and a `VmError::GasDepletion` is returned.

So yeah, some of the out of gas end up there, others are not